### PR TITLE
Improve accuracy of SUIT CSS regex

### LINKED ISSUE #16: suit regexp is weak
Hi, I know that regexps are expensive but the one for suit is very flaky and weak.

For instance the plugin resets things like 

```css
.Component[aria-hidden="true"] {}
```

which results in specificity issues e.g.:

```css
.Component { 
  font-size: 3em;
}

.Component[aria-hidden="true"] {
  display: none;
}
```

Is compiled to

```css
.Component,
.Component[aria-hidden="true"] {
   font-size: medium;
}

.Component { 
  font-size: 3em;
}

.Component[aria-hidden="true"] {
  display: none;
}
```

and because `.Component[aria-hidden="true"]` is more specific than `.Component` `font-size: 3em` is overridden.

You may want to use `postcss-bem-linter`'s [regexps](https://github.com/postcss/postcss-bem-linter/blob/master/lib/preset-patterns.js).

[We are considering](https://github.com/suitcss/preprocessor/issues/43) to add this plugin to suitcss and this is a blocker.
Let me know if you need some help ✌️

### PR DESCRIPTION
**Pull request by simonsmith**
This adds some extra test cases for SUIT selectors and makes the regex
more robust against edge cases.

```
^\.                  - Match a dot at the start of the string
(?:[a-z0-9]*-)?      - Then look for optional namespace
[A-Z]                - Match a single capital letter at the start of Component name
(?:[a-zA-Z0-9]+)     - Match the rest of the component name...
(?:-[a-zA-Z0-9]+)?$  - Match optional descendant
```

https://regex101.com/r/AA4xaq/3

Fixes #16

cc @giuseppeg

### CODE REVIEW

**Comment by maximkoretskiy** [comment]
I think Everything is OK. @giuseppeg Could U review this PR before I merge it?

File: src/rulesMatcher.es6 [hunk 1]
```diff
@@ -1,10 +1,15 @@
+/**
+ * Demo - https://regex101.com/r/AA4xaq/2
+ */
+const suitRegex = /^\.(?:[a-z0-9]*-)?[A-Z](?:[a-zA-Z0-9]+)\b(?!\[|:|\.is-|--)(?:-[a-zA-Z0-9]+)?/;
```
Comment (giuseppeg) [inline src/rulesMatcher.es6]: `(?!\[|:|\.is-|--)` is not necessary anymore if you add `$` at the and of the regexp

File: src/rulesMatcher.es6 [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
+
 const matchers = {
   bem({selector}) {
     return !selector.match(/(--|:)/);
   },
 
   suit({selector}) {
-    return !(selector.match(/(--|:)/) || selector.match(/\.is\-/i));
+    return suitRegex.test(selector);
```
Comment (giuseppeg) [inline src/rulesMatcher.es6]: to make this more performant we could do:

```javascript
return selector.charAt(0) === '.' && suitRegex.test(selector);
```

File: src/rulesMatcher.es6 (hunk 1 above)
Comment (giuseppeg) [inline src/rulesMatcher.es6]: remember to update the Demo

File: src/rulesMatcher.es6 (hunk 2 above)
Comment (simonsmith) [inline src/rulesMatcher.es6]: Is there a performance issue here though? It already checks for the `.` in the regex.

File: src/rulesMatcher.es6 (hunk 1 above)
Comment (simonsmith) [inline src/rulesMatcher.es6]: Nice, that works much better

**Comment by simonsmith** [comment]
> to prematurely return if the css doesn't contain any @define directive i.e. there is no Component to reset.

Do you think that's needed here? Would be a breaking change

**Comment by giuseppeg** [comment]
@simonsmith do you mean without having to add a `@define` directive? I always add it anyway but yeah if it is a breaking change better not do it.

File: src/rulesMatcher.es6 (hunk 2 above)
Comment (giuseppeg) [inline src/rulesMatcher.es6]: regexps are slow and this check will run on **any** kind of selector

**Comment by simonsmith** [comment]
True, it could be seen as breaking but if it's fixing a bug I would say it was just a patch. 

Regardless though, I didn't think this plugin should enforce usage of `@define`. The main reason for that to exist is to derive the component name, which [may not be needed](https://github.com/postcss/postcss-bem-linter/issues/88) in the future anyway and this plugin doesn't really care what your component name is, only that it exists.

I also wonder if users of BEM in particular use `bem-linter` as much as SUIT.

**Comment by giuseppeg** [comment]
<blockquote>I didn't think this plugin should enforce usage of `@define`</blockquote>
Agreed. 

Up to you to decide whether to add the extra check before testing agains the regexp – I am for add it. Either ways the branch is good to merge imo.

**Comment by simonsmith** [comment]
I see the logic in avoiding the regex 👍 

Will add it

**Comment by simonsmith** [comment]
Okay, updated.

Please have a final look @maximkoretskiy @giuseppeg :)

**Comment by giuseppeg** [comment]
🚢 🇮🇹  

awesome work! thank you @simonsmith

**Comment by maximkoretskiy** [comment]
Great!
